### ISSUE TITLE
EnvironmentVariable - An illegal reflective access operation

### ISSUE THREAD

**Issue by slawekjaranowski** [junit-pioneer/junit-pioneer#387]
On
```
openjdk version "11.0.9" 2020-10-20
OpenJDK Runtime Environment (build 11.0.9+11)
OpenJDK 64-Bit Server VM (build 11.0.9+11, mixed mode)
```
and
```
openjdk version "15.0.1" 2020-10-20
OpenJDK Runtime Environment (build 15.0.1+9)
OpenJDK 64-Bit Server VM (build 15.0.1+9, mixed mode, sharing)
```

we have

```
[ERROR] WARNING: An illegal reflective access operation has occurred
[ERROR] WARNING: Illegal reflective access by org.junitpioneer.jupiter.EnvironmentVariableUtils (file:/..../.m2/repository/org/junit-pioneer/junit-pioneer/1.1.0/junit-pioneer-1.1.0.jar) to field java.util.Collections$UnmodifiableMap.m
[ERROR] WARNING: Please consider reporting this to the maintainers of org.junitpioneer.jupiter.EnvironmentVariableUtils
[ERROR] WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
[ERROR] WARNING: All illegal access operations will be denied in a future release
```

It is **warning** message - tests **pass** as expected

**Comment by nipafx** [comment]
Hi @slawekjaranowski,

I assume you're using Pioneer's `@ClearEnvironmentVariable` or `@SetEnvironmentVariable`, correct? If so, please see [the warning in the documentation](https://junit-pioneer.org/docs/environment-variables/):

> Java considers environment variables to be immutable, so this extension uses reflection to change them. This requires that the `SecurityManager` allows modifications and can potentially break on different operating systems and Java versions. Be aware that this is a fragile solution and consider finding a better one for your specific situation. If you’re running on Java 9 or later, you may have to add `--add-opens=java.base/java.util=ALL-UNNAMED` to your test execution to prevent warnings or even errors.

The warning you describe is the result of the reflective operation that changes environment variables. As far as I know, there's no way for Pioneer to suppress these warnings and even if there were, I would oppose implementing it because this is a fragile operation that deserves a warning 😉 , so I'm inclined to close with issue as _works as intended_.

If you're not using these annotations or otherwise disagree, let me know and we can follow up.

**Comment by slawekjaranowski** [comment]
You are right - thanks for explanation.

In order to set this settings in tests I added to my maven project:


```xml
        <profile>
            <id>jmv-9+</id>
            <activation>
                <jdk>[9,)</jdk>
            </activation>
            <properties>
                <argLine>--add-opens=java.base/java.util=ALL-UNNAMED</argLine>
            </properties>
        </profile>
```

**Comment by nipafx** [comment]
I want to reopen the issue to discuss whether we're already doing our best to inform users of what's going on. At the moment, there's the paragraph in the documentation that I quoted above and we publish a report entry saying "This extension uses reflection to mutate JDK-internal state, which is fragile. Check the Javadoc or documentation for more details."

Additionally, we could:

* print that warning to `System.out` or `System.err`
* include the specific warning in the documentation, so search engines lead users to it

Anything else?

**Comment by Michael1993** [comment]
Report entries get processed by dedicated TestExecutionListeners. I think IntelliJ uses one internally, but other than that I think they are mostly lost in the void. Printing on `System.err` or `System.out` is a bit more 'invasive', but I think it is warranted.

I think the documentation should include the warning as well - using the extension is easy (which is good!) and I can see how that could lead to frustrating situations when/if it breaks. Imagine if the issue needed a more robust solution but was temporarily the extension was "good enough" - then it breaks and now people are scrambling to replace it under a lot more pressure...

**Comment by nipafx** [comment]
We just discussed this on the call and came to the following conclusions:

## Log warning

At the moment, we print a related warning via report entries, but those are often lost in the wild, so an additional warning to `System.xxx` seems to be a good idea. (This could be made configurable, but we don't want to go into that here, so we keep it simple.)

We should print to the same stream (`err` or `out`) that the module system uses for its warning and try to get ours as close to it as reasonable.

## Expand documentation

We want to expand the documentation on this by including the exact error message and possibly solutions (e.g. with Maven) as well.

**Comment by nipafx** [comment]
We've just released 1.2.0, which logs the warning we issue regarding this on `System.out` in addition to JUnit's report entries. More importantly, we've improved the documentation, so it contains [a longer section on this](https://junit-pioneer.org/docs/environment-variables/#warnings-for-reflective-access) including the error message and fixes.

@slawekjaranowski Would love to hear your feedback on the documentation change. Assuming Google would have led you there, would that have helped you out?
